internal/interceptor_header.go

```diff
@@ -0,0 +1,134 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"context"
+	"fmt"
+
+	commonpb "go.temporal.io/api/common/v1"
+)
+
+type headerKey struct{}
+
+// Header provides Temporal header information from the context for reading or
+// writing during specific interceptor calls. See documentation in the
+// interceptor package for more details.
+func Header(ctx context.Context) map[string]*commonpb.Payload {
+	m, _ := ctx.Value(headerKey{}).(map[string]*commonpb.Payload)
+	return m
+}
+
+func contextWithNewHeader(ctx context.Context) context.Context {
+	return context.WithValue(ctx, headerKey{}, map[string]*commonpb.Payload{})
+}
+
+func contextWithoutHeader(ctx context.Context) context.Context {
+	return context.WithValue(ctx, headerKey{}, nil)
+}
+
+func contextWithHeaderPropagated(
+	ctx context.Context,
+	header *commonpb.Header,
+	ctxProps []ContextPropagator,
+) (context.Context, error) {
+	if header == nil {
+		header = &commonpb.Header{}
+	}
+	if header.Fields == nil {
+		header.Fields = map[string]*commonpb.Payload{}
+	}
+	reader := NewHeaderReader(header)
+	for _, ctxProp := range ctxProps {
+		var err error
+		if ctx, err = ctxProp.Extract(ctx, reader); err != nil {
+			return nil, fmt.Errorf("failed propagating header: %w", err)
+		}
+	}
+	return context.WithValue(ctx, headerKey{}, header.Fields), nil
+}
+
+func headerPropagated(ctx context.Context, ctxProps []ContextPropagator) (*commonpb.Header, error) {
+	header := &commonpb.Header{Fields: Header(ctx)}
+	if header.Fields == nil {
+		return nil, fmt.Errorf("context missing header")
+	}
+	writer := NewHeaderWriter(header)
+	for _, ctxProp := range ctxProps {
+		if err := ctxProp.Inject(ctx, writer); err != nil {
+			return nil, fmt.Errorf("failed propagating header: %w", err)
+		}
+	}
+	return header, nil
+}
+
+// WorkflowHeader provides Temporal header information from the workflow context
+// for reading or writing during specific interceptor calls. See documentation
+// in the interceptor package for more details.
+func WorkflowHeader(ctx Context) map[string]*commonpb.Payload {
+	m, _ := ctx.Value(headerKey{}).(map[string]*commonpb.Payload)
+	return m
+}
+
+func workflowContextWithNewHeader(ctx Context) Context {
+	return WithValue(ctx, headerKey{}, map[string]*commonpb.Payload{})
```

Comment:
can use a const in place of headerKey{}?

Reply:
Why? Using strings as context keys is bad practice and we need to stop doing that in other places.

From https://pkg.go.dev/context#WithValue:

> The provided key must be comparable and should not be of type string or any other built-in type to avoid collisions between packages using context. 

From https://go.dev/blog/context:

> To avoid key collisions, userip defines an unexported type key and uses a value of this type as the context key

There are plenty of other articles on this and Golint even warns on it IIRC.

Reply:
Not a string. I was referring to const headerKey = headerKey{}, but since it is a value type, so I guess it does not matter.